Add ble support b u585i

javad123javad opened the pull request:
This patch introduces the Bluetooth Low Energy (BLE) feature to the board.
The board utilizes the STM32WB5MMG as the BLE module. However, As there
was no BLE controller available for this module. Therefore, a board
support package has been added to enable the STM32WB5MMG module to act as
a BLE controller. This is achieved by running Zephyr's hci_uart example on
the STM32WB5MMG module which enables communication with the main
microcontroller over the H:4 HCI transport protocol. So, users must first
build the BLE controller for the BLE module and upload it via on board
ST-Link,then they can uses Zephyr Bluetooth demos on the development board
Note that there was no hardware flow control wiring available on the
board, which is why it has been disabled in the both main board and BLE
module Device Tree. As the board doesn't support HW flow control, users
must must set CONFIG_BT_HCI_ACL_FLOW_CONTROL=n in project files.

Commit:
boards: arm: stm32wb5mmg: add Bluetooth support

This patch introduces the Bluetooth Low Energy (BLE) feature to the board.
The board utilizes the STM32WB5MMG as the BLE module. However, As there
was no BLE controller available for this module. Therefore, a board
support package has been added to enable the STM32WB5MMG module to act as
a BLE controller. This is achieved by running Zephyr's hci_uart example on
the STM32WB5MMG module which enables communication with the main
microcontroller over the H:4 HCI transport protocol. So, users must first
build the BLE controller for the BLE module and upload it via on board
ST-Link,then they can uses Zephyr Bluetooth demos on the development board
Note that there was no hardware flow control wiring available on the
board, which is why it has been disabled in the both main board and BLE
module Device Tree.

Signed-off-by: Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>

arnout commented on boards/arm/b_u585i_iot02a/doc/index.rst:
```diff
@@ -195,6 +195,8 @@ The Zephyr b_u585i_iot02a board configuration supports the following hardware fe
 +-----------+------------+-------------------------------------+
 | AES       | on-chip    | crypto                              |
 +-----------+------------+-------------------------------------+
+| RADIO     | on-chip    | Bluetooth Low Energy (BLE)          |
```
The BLE is not on-chip, so I think the controller column should specify STM32WB5MMG instead.

javad123javad replied:
Thanks, I will fix it.

arnout commented on boards/arm/stm32wb5mmg/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# STM32WB5MMG board configuration
+
+# Copyright (c) 2019 Linaro Limited
```
Is this copyright year and author correct?

arnout commented on boards/arm/stm32wb5mmg/Kconfig.defconfig:
```diff
@@ -0,0 +1,16 @@
+# STM32WB5MMG Bluetooth module board configuration
+
+# Copyright (c) 2019 Linaro Limited
```
Same here, is this copyright year and author correct?

arnout commented on boards/arm/stm32wb5mmg/stm32wb5mmg.dts:
```diff
@@ -0,0 +1,209 @@
+};
+
+&i2c1 {
+	pinctrl-0 = <&i2c1_scl_pb8 &i2c1_sda_pb9>;
```
Pin mapping and peripherals in stm32wb5mmg.rst don't mention I2C...

javad123javad replied:
i2c1 is not connected to any peripheral. So I will remove it from the device three.

erwango replied:
Ok, please do the same for the whole module configuration and update the doc accordingly.
It should be clear when reading the doc what are the resources required to for module to operate properly

erwango commented on boards/arm/b_u585i_iot02a/doc/index.rst:
```diff
@@ -195,7 +195,7 @@ The Zephyr b_u585i_iot02a board configuration supports the following hardware fe
 | AES       | on-chip    | crypto                              |
 +-----------+------------+-------------------------------------+
-| RADIO     | on-chip    | Bluetooth Low Energy (BLE)          |
+| RADIO     | STM32WB5MMG| Bluetooth Low Energy (BLE)          |
```
Please do any changes in the faulty commits directly. We don't merge fix commits

erwango commented on boards/arm/stm32wb5mmg/Kconfig.board:
```diff
@@ -1,6 +1,6 @@
 # STM32WB5MMG board configuration
 
-# Copyright (c) 2019 Linaro Limited
+# Copyright (c) 2023 Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
```
Ditto

erwango commented on boards/arm/stm32wb5mmg/Kconfig.defconfig:
```diff
@@ -1,6 +1,6 @@
 # STM32WB5MMG Bluetooth module board configuration
 
-# Copyright (c) 2019 Linaro Limited
```
Ditto

erwango commented on boards/arm/stm32wb5mmg/doc/stm32wb5mmg.rst:
```diff
@@ -0,0 +1,303 @@
+========
+
+STM32WB5MMG needs an external ST-Link for flashing the firmware. If you are using
+one of the  ``b_u585i_iot02a`` or ``nucleo_wb55rg`` boards, you can use the on-board
```
This is not true for `nucleo_wb55rg`, which uses a STM32WB55RC SoC and not a module. Please focus documentation on `b_u585i_iot02a` case.

javad123javad replied:
`nucleo_wb55rg` removed from the document. Now the focus is only on the `b_u585i_iot02a` board.

erwango commented on boards/arm/stm32wb5mmg/doc/stm32wb5mmg.rst:
```diff
@@ -0,0 +1,303 @@
+one of the  ``b_u585i_iot02a`` or ``nucleo_wb55rg`` boards, you can use the on-board
+ST-Link to flash the  STM32WB5MMG module.
+
+If you prefer, you can use pyocd, but it requires to enable "pack" support with
+the following pyocd command:
```
Don't mix probes and flashing tools.
Pyocd is not an alternative to ST-Link. It is an alternative to openocd or STM32CubeProgrammer.

erwango commented on boards/arm/stm32wb5mmg/doc/stm32wb5mmg.rst:
```diff
@@ -0,0 +1,303 @@
+These binaries are delivered in STM32WB Cube packages, under
+Projects/STM32WB_Copro_Wireless_Binaries/STM32WB5x/
+For compatibility information with the various versions of these binaries,
+please check `modules/hal/stm32/lib/stm32wb/hci/README <https://github.com/zephyrproject-rtos/hal_stm32/blob/main/lib/stm32wb/hci/README>`__
```
Put the link at the bottom of the file and use a reference here.

erwango commented on boards/arm/stm32wb5mmg/doc/stm32wb5mmg.rst:
```diff
@@ -0,0 +1,303 @@
+Flashing `hci_uart` application to STM32WB5MMG
+----------------------------------------
+
+Connect the B-U585I-IOT02A to your host computer using the USB port. Put
```
This isn't quite clear vs the previous explanation ("STM32WB5MMG needs an external ST-Link"). Maybe you could clarify ?

javad123javad replied:
Thanks, I have rewritten the explanation with the focus on `B-U585I-IOT02A` board, such that users can flash the BLE module on this board.

erwango reviewed (changes requested):
Some changes required mainly around doc and device tree.
Otherwise, it would be required to split in 2 commits:
- new board introduction
- b_u585i_iot update

Thanks !

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts:
```diff
@@ -17,6 +17,10 @@
 		zephyr,flash = &flash0;
 		zephyr,code-partition = &slot0_partition;
+		zephyr,bt-uart=&uart4;
+		zephyr,bt-c2h-uart=&uart4;
```
I would rather see that defined in the module part

javad123javad replied:
> I would rather see that defined in the module part

Not sure I understand this correctly. Because this configuration is related to the main microcontroller (U585i), and the UART4 is only connected to the BLE module via this link. If we put it in module device three, how we can address the main application how find the proper UART port for BLE communications?

erwango replied:
This comment applied to `zephyr,bt-c2h-uart`. Sorry, it was misleading.
```
   * - zephyr,bt-c2h-uart
     - Selects the UART used for host communication in the
       :ref:`bluetooth-hci-uart-sample`
```
This sample applies to the module, not the host processor.

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts:
```diff
@@ -17,6 +17,10 @@
 		zephyr,flash = &flash0;
 		zephyr,code-partition = &slot0_partition;
+		zephyr,bt-uart=&uart4;
+		zephyr,bt-c2h-uart=&uart4;
+		zephyr,bt-mon-uart=&uart4;
```
Can you confirm using btmon is working with this setting ?

javad123javad replied:
Not confirmed. removed from the device tree.

Commit:
boards: arm: b_u585i_iot02a: add bluetooth support

This commits makes it possible to use the onboard bluetooth
module (STM32WB5MMG) with existing zephyr bluetooth samples.
Note that there was no hardware flow control wiring
available on the board, which is why it has been disabled
in the both main board and BLE module Device Tree. As the
board doesn't support HW flow control, users must set
CONFIG_BT_HCI_ACL_FLOW_CONTROL=n in project files.

Signed-off-by: Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>

erwango commented on boards/arm/stm32wb5mmg/stm32wb5mmg_defconfig:
```diff
@@ -0,0 +1,27 @@
+CONFIG_PINCTRL=y
+
+# LSE defined as LPTIM clock source by the DTS
+CONFIG_STM32_LPTIM_CLOCK_LSE=y
```
No longer required, see https://github.com/zephyrproject-rtos/zephyr/pull/66714

javad123javad commented on boards/arm/stm32wb5mmg/stm32wb5mmg.dts:
```diff
@@ -0,0 +1,190 @@
+};
+
+
+&rtc {
```
I have removed the following configs:

- timers1, 
- timers2,
- lptim1,
- rng,
- vbat,
- clk_lsi1,
- clk_lse,
- clk48
- rtc

erwango commented on boards/arm/stm32wb5mmg/stm32wb5mmg.dts:
```diff
@@ -0,0 +1,144 @@
+		#size-cells = <1>;
+
+		/*
+		 * Configure partitions while leaving space for M0 BLE f/w
+		 * Since STM32WBCube release V1.13.2, only _HCIOnly_ f/w are supported.
+		 * These FW are expected to be located not before 0x080DB000
+		 * Current partition is using the first 876K of the flash for M4
+		 */
+
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x00000000 DT_SIZE_K(48)>;
+		};
+		slot0_partition: partition@c000 {
+			label = "image-0";
+			reg = <0x0000c000 DT_SIZE_K(400)>;
+		};
+		slot1_partition: partition@70000 {
+			label = "image-1";
+			reg = <0x00070000 DT_SIZE_K(400)>;
+		};
+		scratch_partition: partition@d4000 {
+			label = "image-scratch";
+			reg = <0x000d4000 DT_SIZE_K(16)>;
+		};
+		storage_partition: partition@d8000 {
+			label = "storage";
+			reg = <0x000d8000 DT_SIZE_K(8)>;
+		};
+
+	};
+};
+
+&vref {
+	status = "okay";
+};
+
+&vbat {
+	status = "okay";
+};
```
Do you actually need those ?

javad123javad replied:
Flash layout simplified to :
```
&flash0 {
        partitions {
                compatible = "fixed-partitions";
                #address-cells = <1>;
                #size-cells = <1>;

                /*
                 * Configure partitions while leaving space for M0 BLE f/w
                 * Since STM32WBCube release V1.13.2, only _HCIOnly_ f/w are supported.
                 * These FW are expected to be located not before 0x080DB000
                 * Current partition is using the first 876K of the flash for M4
                 */

                boot_partition: partition@0 {
                        label = "mcuboot";
                        reg = <0x00000000 DT_SIZE_K(48)>;
                };

                slot0_partition: partition@c000 {
                        label = "image-0";
                        reg = <0x0000c000 DT_SIZE_K(400)>;
                };
```
and `vref`,`vbat` and `aes` have been removed. When trying to remove `adc` some errors were reported. I holt `iwdg` because I was not sure whether it was used by HCI core internally or not. If not required, I can also remove it.

erwango commented on boards/arm/stm32wb5mmg/doc/stm32wb5mmg.rst:
```diff
@@ -0,0 +1,304 @@
+B-U585I-IOT02A board:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/bluetooth/observer
```
I had a success with this sample but with peripheral_hr, I got the following issue:
```
Bluetooth init failed (err -22)
```
Did you observe this as well ?

javad123javad replied:
That's because the connenction between BLE module and main micro controller doesn't have HW Flow control wiring. So, as it has been mentioned in the commit log, users should disable it with by adding `CONFIG_BT_HCI_ACL_FLOW_CONTROL=n` to prj.conf files. 

> commit 05d12f47f4c5757276d53d2518dad516ca4f93d1 (HEAD -> add_ble_support_b_u585i, origin/add_ble_support_b_u585i)
> Author: Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
> Date:   Mon Jan 15 15:01:40 2024 +0100
> 
>     boards: arm: b_u585i_iot02a: add bluetooth support
>     
>     This commits makes it possible to use the onboard bluetooth
>     module (STM32WB5MMG) with existing zephyr bluetooth samples.
>     Note that there was no hardware flow control wiring
>     available on the board, which is why it has been disabled
>     in the both main board and BLE module Device Tree. As the
>     board doesn't support HW flow control, users must set
>     CONFIG_BT_HCI_ACL_FLOW_CONTROL=n in project files.
console output for `peripheral_hr` after disabling `CONFIG_BT_HCI_ACL_FLOW_CONTROL` :
```
*** Booting Zephyr OS build zephyr-v3.5.0-4217-g05d12f47f4c5 ***
[00:00:00.051,000] <inf> bt_hci_core: Identity: 02:80:E1:00:00:00 (public)
[00:00:00.052,000] <inf> bt_hci_core: HCI: version 1.0b (0x00) revision 0xa077, manufacturer 0x0030
[00:00:00.052,000] <inf> bt_hci_core: LMP: version 1.0b (0x00) subver 0x2177
Bluetooth initialized
Advertising successfully started

```

erwango commented on boards/arm/stm32wb5mmg/stm32wb5mmg.dts:
```diff
@@ -0,0 +1,144 @@
+	status = "okay";
+};
+
+&clk_hse {
+	status = "okay";
+};
+
+&clk_lse {
+	status = "okay";
+};
+
+&clk_hsi48 {
+	status = "okay";
+};
```
Do you actually need all these clocks ?

javad123javad replied:
For the `hsi48` when it is disabled, the following warning is reported. So, I put it in the DT:
```
ipm_stm32wb.c:510:2: warning: #warning Bluetooth IPM requires HSI48 clock to be enabled using device tree [-Wcpp]
  510 | #warning Bluetooth IPM requires HSI48 clock to be enabled using device tree
```

When `clk_lse` is disabled, the following error is reported:
```
ASSERTION FAIL [err == 0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:331
	command opcode 0x0c03 timeout with err -11
[00:00:10.000,000] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000000  r2/a3:  0x00000002
[00:00:10.000,000] <err> os: r3/a4:  0x200005f8 r12/ip:  0x00002710 r14/lr:  0x0800335b
[00:00:10.000,000] <err> os:  xpsr:  0x49000000
[00:00:10.000,000] <err> os: Faulting instruction address (r15/pc): 0x08003366
[00:00:10.000,000] <err> os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:00:10.000,000] <err> os: Current thread: 0x200010d0 (unknown)
[00:00:10.056,000] <err> os: Halting system
```
Finally, when `clk_hse` is disabled, the following error is reported:
```
[00:00:00.000,000] <err> bt_driver: Unknown H:4 type 0x00
ASSERTION FAIL [err == 0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:331
	command opcode 0x0c03 timeout with err -11
[00:00:10.000,000] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000000  r2/a3:  0x00000002
[00:00:10.000,000] <err> os: r3/a4:  0x200005f8 r12/ip:  0x00002710 r14/lr:  0x0800335b
[00:00:10.000,000] <err> os:  xpsr:  0x49000000
[00:00:10.000,000] <err> os: Faulting instruction address (r15/pc): 0x08003366
[00:00:10.000,000] <err> os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:00:10.000,000] <err> os: Current thread: 0x200010d0 (unknown)
[00:00:10.056,000] <err> os: Halting system

```

erwango replied:
Ok, let's keep them then ;). Thanks.

erwango commented:
@javad123javad Please fix compliance issue

javad123javad commented:
> @javad123javad Please fix compliance issue

Sure, As it is approved, what is the proper way to do it? should I rebase to my commit or to the base commit?
You explained it [here](https://github.com/zephyrproject-rtos/zephyr/pull/67819#issuecomment-1903433907), but I didn't understand it correctly.

erwango commented:
> Sure, As it is approved, what is the proper way to do it? should I rebase to my commit or to the base commit?

Yep: fix and force-push. Avoid rebase if not required. You'll lost approval, but we should be able to identify change easily and re-approve quickly.
